src/Oscoin/Crypto/Address/Serialisation.hs

```diff
@@ -154,43 +122,39 @@ deserializeAddressType = do
 -- | Deserializes the 'AddressFormat'.
 deserializeAddressFormat :: Decoder AddressFormat
 deserializeAddressFormat = do
-    tag <- decodeWord8
+    tag <- withExceptT DecodeError parseWord8
     case tag of
       0x01 -> pure $ AddressFormat AddressFormatTag_CBOR
       _    -> throwError $ UnknownAddressFormat tag
 
 -- | Deserializes the 'AddressPayload' by piggybacking on CBOR.
 deserializeAddressPayload
-    :: ( ByteArrayAccess (ShortHash c)
-       , CBOR.Serialise (ShortHash c)
-       )
+    :: CBOR.Serialise (PublicKey c)
     => AddressFormat
     -> Decoder (AddressPayload c)
 deserializeAddressPayload (AddressFormat tag) = do
     binaryBlob <- get
     -- Extension point to support multiple payloads in the future.
     case tag of
       AddressFormatTag_CBOR ->
-          lift . liftEither $
+          liftEither $
               first DeserializePayloadError
                     (CBOR.deserialiseOrFail (toS binaryBlob))
 
 {------------------------------------------------------------------------------
   Serialisation
 ------------------------------------------------------------------------------}
 
-instance ( ByteArrayAccess (ShortHash c)
-         , CBOR.Serialise (ShortHash c)
-         ) => CBOR.Serialise (AddressPayload c) where
-    encode (AddressPayload (Tagged p)) =
+instance CBOR.Serialise (PublicKey c) => CBOR.Serialise (AddressPayload c) where
+    encode (AddressPayload_V0 pk) =
            CBOR.encodeListLen 2
         <> CBOR.encode (0 :: Word8)
-        <> CBOR.encodeByteArray (toSliced p)
+        <> CBOR.encode pk
```

Comment:
I think this is what you were talking about - but we should really ensure that whatever this does is consistent with some hard-coded value, eg. that the tags used by cbor are the same accross all versions, or hand-roll the instance of PK (maybe we do this already)?

Reply:
I think that trying to be consistent with CBOR tags across all versions is a nightmare and simply not feasible 😅 😁 . Yes, we can (and should) hard-roll the instance of PK, but I think I am hinting  at more general use of a pattern here: whenever we have types that might evolve, we should probably represent them as sum types or GADTs (a bit like I did here for the `AddressPayload` or you guys did for the `Tx`) so that we can match the particular constructor with the particular serialised representation. This is probably obvious for the most parts but I guess we should start thinking which are those types and represent them consistently (and perhaps dump some documentation somewhere on how to write new types in the future).

Reply:
Hmm I'm not sure how the sum types help here - we still have to have a spec somewhere of what the different tags mean.

Reply:
By all means, one doesn't exclude the other -- I was merely pointing out that you do need a sum type (or GADT) if you don't want to have each version bump to cause a cascade effect everywhere else (unless that's the intended side-effect, of course), while being able to map each tag to a different constructor.

Ideally we could also think about using some SOP/TH dark magic to generate such documentation directly from the code.

Reply:
Ah ok, I'm more concerned about 3rd party clients being able to decode the format reliably, but yes, the sum-type is also important.